File: src/main/java/cn/taskeren/minequery/callback/NotHit.java
package cn.taskeren.minequery.callback;

import net.fabricmc.fabric.api.event.player.AttackEntityCallback;
import net.minecraft.entity.Entity;
import net.minecraft.entity.passive.IronGolemEntity;
import net.minecraft.entity.passive.VillagerEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.EntityHitResult;
import net.minecraft.world.World;
import org.jetbrains.annotations.Nullable;

import static cn.taskeren.minequery.MineQueryMod.config;

public class NotHit {

	private NotHit() {}

	public static void register() {
		AttackEntityCallback.EVENT.register(IronGolem::interact);
		AttackEntityCallback.EVENT.register(Villager::interact);
	}

	public static class IronGolem {

		private IronGolem() {}

		public enum PreventDamage {
			ALL,
			NATURAL_ONLY,
			PLAYER_ONLY,
			NONE
		}

		public static ActionResult interact(PlayerEntity player, World world, Hand hand, Entity entity, @Nullable EntityHitResult hitResult) {
			if(!config().notHitConfig.notHit) {
				return ActionResult.PASS;
			}
			if(entity instanceof IronGolemEntity) {
				boolean playerCreated = ((IronGolemEntity) entity).isPlayerCreated();
				switch(config().notHitConfig.ironGolem) {
					case ALL -> {
						return ActionResult.FAIL;
					}
					case NATURAL_ONLY -> {
						return playerCreated ? ActionResult.PASS : ActionResult.FAIL;
					}
					case PLAYER_ONLY -> {
						return playerCreated ? ActionResult.FAIL : ActionResult.PASS;
					}
					case NONE -> {
						return ActionResult.PASS;
					}
				}
			}
			return ActionResult.PASS;
		}
	}

	public static class Villager {

		private Villager() {}
		public static ActionResult interact(PlayerEntity player, World world, Hand hand, Entity entity, @Nullable EntityHitResult hitResult) {
			if(!config().notHitConfig.notHit) {
				return ActionResult.PASS;
			}
			if(entity instanceof VillagerEntity) {
				return config().notHitConfig.villager ? ActionResult.FAIL : ActionResult.PASS;
			}
			return ActionResult.PASS;
		}
	}

}

File: src/main/kotlin/cn/taskeren/minequery/features/NotHit.kt
package cn.taskeren.minequery.features

import cn.taskeren.minequery.MineQuery
import cn.taskeren.minequery.config.PreventDamageIronGolem
import net.fabricmc.fabric.api.event.player.AttackEntityCallback
import net.minecraft.entity.passive.IronGolemEntity
import net.minecraft.entity.passive.VillagerEntity
import net.minecraft.util.ActionResult

object NotHit {

	private val c get() = MineQuery.config.notHitConfig

	fun init() {

		AttackEntityCallback.EVENT.register { _, _, _, entity, _ ->
			if(!c.notHit) {
				return@register ActionResult.PASS
			}

			if(entity is IronGolemEntity) {
				val isPlayerCreated = entity.isPlayerCreated
				when(c.ironGolem) {
					PreventDamageIronGolem.ALL -> {
						return@register ActionResult.FAIL
					}
					PreventDamageIronGolem.NATURAL_ONLY -> {
						if(!isPlayerCreated) {
							return@register ActionResult.FAIL
						}
					}
					PreventDamageIronGolem.PLAYER_ONLY -> {
						if(isPlayerCreated) {
							return@register ActionResult.FAIL
						}
					}
					PreventDamageIronGolem.NONE -> {
						return@register ActionResult.PASS
					}
				}
			}

			if(entity is VillagerEntity) {
				if(c.villager) {
					return@register ActionResult.FAIL
				}
			}

			ActionResult.PASS
		}

	}

}